src/app/components/animate/animate.ts

```diff
@@ -0,0 +1,178 @@
+import { NgModule, Directive, ElementRef, Input, Renderer2, Inject } from '@angular/core';
+import { CommonModule, DOCUMENT } from '@angular/common';
+import { DomHandler } from '../dom/domhandler';
+
+@Directive({
+    selector: '[pAnimate]',
+    host: {
+        '[class.p-animate]': 'true'
+    }
+})
+export class Animate {
+
+    @Input('parent') parentElement: any;
+
+    @Input() enterClass: string;
+
+    @Input() leaveClass: string;
+
+    documentScrollListener: Function | null = null;
+
+    loadListener: Function = () => { };
+
+    entered: boolean;
+
+    observer: IntersectionObserver;
+
+    loaded: boolean;
+
+    scroll: boolean;
+
+    constructor(@Inject(DOCUMENT) private document: Document, private host: ElementRef, public el: ElementRef, public renderer: Renderer2) { }
+
+    ngOnInit() {
+        if (this.isImage()) {
+            if (this.isInViewport()) {
+                this.enter();
+            }
+            this.bindLoadListener();
+        }
+    }
+
+    constructIntersectionObserver() {
+        const options = {
+            root: null,
+            rootMargin: '0px',
+            treshold: 1.0
+        }
+
+        this.observer = new IntersectionObserver((el) => this.isVisible(el), options);
+        this.observer.observe(this.host.nativeElement);
+    }
+
+    isVisible(element) {
```

Comment:
I think this code should run out of ngzone (since IntersectionObserver can throw events frequently).